### PR TITLE
[SYCLomatic] Emit warning if kernel function pointer cannot be deduced

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 1]
```diff
@@ -5442,9 +5446,13 @@ std::shared_ptr<KernelCallExpr> KernelCallExpr::buildFromCudaLaunchKernel(
         Kernel->ArgsInfo.emplace_back(Parm, ArgsArray, Kernel.get());
       }
     }
-    return Kernel;
+  } else {
+    DiagnosticsUtils::report(LocInfo.first, LocInfo.second,
+                             Diagnostics::UNDEDUCED_KERNEL_FUNCTION_POINTER,
+                             true, false, ExprAnalysis::ref(CE->getArg(0)));
+    Kernel->buildCalleeInfo(CE->getArg(0));
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]:
```suggestion
    Kernel->buildCalleeInfo(CE->getArg(0));
    DiagnosticsUtils::report(LocInfo.first, LocInfo.second,
                             Diagnostics::UNDEDUCED_KERNEL_FUNCTION_POINTER,
                             true, false, Kernel->getName());
```

**Review by ziranzha — APPROVED** [review]

File: clang/test/dpct/launch-kernel-cooperative-usm.cu [hunk 2]
```diff
@@ -83,9 +83,13 @@ int main() {
   void *kernel_func = (void *)&kernel;
 
   // CHECK: /*
-  // CHECK-NEXT: DPCT1007:{{[0-9]+}}: Migration of cudaLaunchCooperativeKernel is not supported.
+  // CHECK-NEXT: DPCT1123:{{[0-9]+}}: The kernel function pointer cannot be deduced correctly. You need cast "kernel_func" to a function pointer, and may adjust the kernel function argument(s) and the dimension of the sycl::nd_item type manually.
```
Comment (AndyCHHuang) [inline clang/test/dpct/launch-kernel-cooperative-usm.cu]: The kernel function pointer cannot be deduced correctly. You need cast "kernel_func" to a function pointer. 
According to the kernel function definition, adjusting the kernel function arguments and the dimension of the sycl::nd_item may be also required.

**Review by AndyCHHuang — APPROVED** [review]
LGTM

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**